### DISCUSSION TITLE
Realtime updater not starting

### DISCUSSION THREAD

**Discussion by gino-ats** [mia-platform/community#297] [Q&A] [answered]
### Which section?

Fast Data

### Which is the version?

 11.0.2

### Description

We cloned a working fast-data project with 3 realtime updaters to create a mirror MongoDB with exactly the same data.  The general idea is to use same ingestion topic (different kafka-group-id) and sink data into a new MongoDB server.

For deployment reasons we need to differentiate microservices names even between different projects thus during cloning we renamed the projection with "-1" suffix (projection A became A-1 and auto-generated microservices are fast-data-A-1-realtime-updater).   

But the problem is the new realtime updaters (all of them) fail to start, and after a lot of comparison with original version, wd can't figure out why:

Stack trace:

> real-time-updater@7.5.4 start
> lc39 index.js --port=3000 --log-level=debug --prefix=/
(node:18) [FSTDEP006] FastifyDeprecation: You are decorating Request/Reply with a reference type. This reference is shared amongst all requests. Use onRequest hook instead. Property: httpClientMetrics
(Use `node --trace-warnings ...` to show where the warning was created)
{"level":20,"time":1690555280786,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","invariantTopicMap":"{\"Aaaaaa\":\"eeee-fast-data-Aaaaaa\",\"Bbbbbb\":\"eeee-fast-data-Bbbbbb\"}","msg":"invariant topic map environment value"}
{"level":20,"time":1690555280787,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","castFunctionsFolder":"/home/node/app/configurations/castFunctionsFolder","mapTableFolder":"/home/node/app/configurations/mapTableFolder","strategiesFolder":"/home/node/app/configurations/strategies","kafkaProjectionChangesFolder":"/home/node/app/configurations/kafkaProjectionChangesFolder","kafkaProjectionUpdatesFolder":"/home/node/app/configurations/kafkaProjectionUpdatesFolder"}
{"level":20,"time":1690555280787,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","topicsList":["eeee-fast-data-Aaaaaa","eeee-fast-data-Bbbbbb"],"msg":"topics list"}
{"level":20,"time":1690555280795,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","castFunctions":["castToDateOrNull","castToLowercaseString","castToTimestampOrNull","defaultCastStringToBoolean","defaultCastToArrayOfObject","defaultCastToDate","defaultCastToFloat","defaultCastToInteger","defaultCastToObject","defaultCastToString","defaultCastUnitTimestampToISOString","defaultIdentity"],"msg":"castFunction"}
{"level":20,"time":1690555280799,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","filePath":"/home/node/app/configurations/kafkaProjectionChangesFolder/kafkaProjectionChanges.json","msg":"no Kafka Projection Changes file has been set"}
{"level":20,"time":1690555280800,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","filePath":"/home/node/app/configurations/kafkaProjectionUpdatesFolder/kafkaProjectionUpdates.json","msg":"no Kafka Projection Changes file has been set"}
{"level":20,"time":1690555280800,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","kafkaProjectionUpdates":{},"kafkaProjectionChanges":{},"msg":"received updates and changes"}
{"level":20,"time":1690555280804,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","files":["erSchema.json"],"msg":"files found in er schema folder"}
{"level":20,"time":1690555280910,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","neededFieldsMap":{"Bbbbbb->Xxxxx":["id"],"Xxxxx->Yyyyy":["uuid"],"{"level":20,"time":1690555280910,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","neededFieldsMap":{"Bbbbbb->Xxxxx":["id"],"Xxxxx->Yyyyy":["uuid"],"Yyyyy->sv-aaaaaaaa":["PROC_INST_ID_"]},"msg":"calculated needed fields map for automatic strategy execution"}
...
{"level":20,"time":1690555280911,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","projectionName":"Aaaaaa","singleView":"sv-aaaaaaaa","msg":"use automatic strategies"}
{"level":20,"time":1690555280912,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","projectionName":"Bbbbbb","singleView":"sv-aaaaaaaa","msg":"use automatic strategies"}
...
{"level":20,"time":1690555280912,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","removed":[],"msg":"list of topics or invariant removed from mapTable because unused"}
{"level":20,"time":1690555280913,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","mapTable":["Aaaaaa","Bbbbbb"],"msg":"mapTable"}
{"level":30,"time":1690555280992,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Server listening at http://0.0.0.0:3000"}
{"level":50,"time":1690555285444,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","debugInfo":{"status":"KO","name":"real-time-updater","version":"7.5.4"},"msg":"Service status is not ok"}
{"level":50,"time":1690555295437,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","debugInfo":{"status":"KO","name":"real-time-updater","version":"7.5.4"},"msg":"Service status is not ok"}
{"level":50,"time":1690555305443,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","debugInfo":{"status":"KO","name":"real-time-updater","version":"7.5.4"},"msg":"Service status is not ok"}
{"level":50,"time":1690555310955,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","debugInfo":{"status":"KO","name":"real-time-updater","version":"7.5.4"},"msg":"Service status is not ok"}
{"level":50,"time":1690555310956,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","error":{"reason":{"type":"Unknown","servers":{},"stale":false,"compatible":true,"heartbeatFrequencyMS":10000,"localThresholdMS":15,"setName":null,"maxElectionId":null,"maxSetVersion":null,"commonWireVersion":0,"logicalSessionTimeoutMinutes":null}},"msg":"Projections storage canNOT startup"}
{"level":60,"time":1690555310956,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","error":"Projections storage canNOT startup","msg":"importer was stopped by error"}
{"level":50,"time":1690555310957,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Consumer is not connected to any topic"}
{"level":40,"time":1690555312958,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Tearing down"}
{"level":40,"time":1690555312958,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Heartbeat stopped"}
{"level":40,"time":1690555312959,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Projections storage teared down"}
{"level":40,"time":1690555312960,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Projections changes storage teared down"}
{"level":40,"time":1690555312963,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"Kafka consumer disconnected"}
{"level":60,"time":1690555312964,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","msg":"closing the server"}
npm notice 
npm notice New minor version of npm available! 9.5.1 -> 9.8.1
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.8.1>
npm notice Run `npm install -g npm@9.8.1` to update!
npm notice 

### Actual Outcome

the realtime updater don't start

**Comment by flower-of-the-bridges** [comment]
Hi @gino-ats !

What is the name of the topics in the new projection you have created? 

At the bottom of the projection page, you can see their names. This problem usually happens when a topic has not been found on a Kafka cluster

So for example, considering this line that you've provided:

```
{"level":20,"time":1690555280787,"pid":18,"hostname":"fast-data-xxxxxx-1-realtime-updater-5bb588b89c-4j2p9","topicsList":["eeee-fast-data-Aaaaaa","eeee-fast-data-Bbbbbb"],"msg":"topics list"}
```
Are the tow topics in the array the same of the kafka cluster?

Bests
Giovanni

**Reply by gino-ats** [reply]
Hi Giovanni,

for sensitive data reasons I can't share real topic names, but they surely exists on kafka cluster since we "cloned" a console project that is working and we do not changed the  topic names.  I also check the kafka address,  it's ok and I think there should be different exception when the cluster is not reachable (ie: wrong IP or typo on name).

thanks
gino

**Reply by flower-of-the-bridges** [reply, marked as answer]
Hi @gino-ats 

Sorry for the misunderstanding, of course, you can't share sensitive information!

I've checked the logs in detail and I see the following error:

```
{"type":"Unknown","servers":{},"stale":false,"compatible":true,"heartbeatFrequencyMS":10000,"localThresholdMS":15,"setName":null,"maxElectionId":null,"maxSetVersion":null,"commonWireVersion":0,"logicalSessionTimeoutMinutes":null}},"msg":"Projections storage canNOT startup"}
```

It seems that the connection to the MongoDB database where projections are stored fails. Are the following variables configured to the microservice?

- `MONGODB_URL`
- `PROJECTIONS_DATABASE_NAME`

 Maybe the value of `MONGODB_URL` was stored in a secret variable, was that also moved on the new project?

Let us know
Giovanni

**Reply by gino-ats** [reply]
Hi Giovanni,

I checked and it seems all is ok (attached you can see the config).  Ad double-check I created a new microservice (a GUI frontend on MongoDB) connected to the same MongoDB instance using same connection string and it works, so it shouldn't be the source of the issue.

Let me known if I can do some other test

Thanks
![1](https://github.com/mia-platform/community/assets/78624327/8ef89e99-fe8b-4677-a82b-6dbe88c82d71)
![2](https://github.com/mia-platform/community/assets/78624327/6568c793-7f04-45ce-b010-bf09764b841a)

**Reply by gino-ats** [reply]
Hi Giovanni,

sorry my test with GUI client was incomplete, I didn't check the frontend navigation and now I see there are issue with MongoDb connection as you stated.

Please consider closing this thread.

Thank you very much for your help.

ciao
gino

**Reply by flower-of-the-bridges** [reply]
Hi @gino-ats 

I'm happy you managed to solve the issue! 😃 

Buona giornata
Giovanni
